### ISSUE TITLE
Make Argon2id The Default PBKDF for LUKS

### ISSUE THREAD

**Issue by Argone2** [QubesOS/qubes-issues#5205]
The problem you're addressing (if any)
Make FDE stronger as well as more post-quantum secure.

Describe the solution you'd like
Make Argon2id encryption the default PBKDF for LUKS, abandoning PBKDF2. Or, at the very least update cryptsetup to a modern version so we can do it ourselves.

Where is the value to a user, and who might that user be?
It will make the user safer from having the contents of their hard drive divulged even in the case of a weak password. Additionally, it will especially make their data more post-quantum secure because quantum computers do not impact the entropy provided by key-stretching algorithms.

Describe alternatives you've considered
Argon2id is superior to argon2i and argon2d as well as the current PBKDF2.

More info
As it stands it does not even seem that the cryptsetup in the dom0 of Qubes even supports argon2id because it's still all the way back at 1.7.5 whereas both my Fedora and Debian qube has the version 2.0.6. This is also evidenced by a test for it not being present for the command cryptsetup benchmark in dom0 as well as lack of speech about it in the manual.

The minor performance penalty should not be a factor considering that Qubes is an OS for people with modern/beefy computers.

***Qubes Version***
4.x

Relevant documentation you've consulted
Whonix recommends switching to Argon2id: https://www.whonix.org/wiki/Passwords#Key-Stretching

**Comment by andrewdavidwong** [comment]
I imagine that this will automatically be taken care of the next time there is a Qubes release that updates dom0 to a sufficiently new Fedora release.

**Comment by Argone2** [comment]
An update to dom0's Fedora would only make the option available for re-encryption after the fact.

Here are the defaults for the up-to-date Fedora qube with Argon algs (cryptsetup --help):
Default PBKDF for LUKS1: pbkdf2, iteration time: 2000 (ms)
Default PBKDF for LUKS2: argon2i

But following what Whonix recommends Argon2id should be used not Argon2i. They are different and you can read more about the differences here: https://crypto.stackexchange.com/questions/48935/why-use-argon2i-or-argon2d-if-argon2id-exists
and here: https://en.wikipedia.org/wiki/Argon2

But in a nutshell (from Wikipedia, condensed):
    Argon2d maximizes resistance to GPU cracking attacks
    Argon2i is optimized to resist side-channel attacks
    Argon2id is a hybrid version. It follows the Argon2i approach for the first pass over memory and the Argon2d approach for subsequent passes.

I believe Argon2id is widely considered superior and that it should be chosen as a default for the encryption upon installation of Qubes.

Also note that read the manual it would seem cryptsetup only supports Argon2i and Argon2id, so one of them would have to be chosen.

BTW: Please do upgrade the version of Fedora on dom0. I've noticed on multiple occasions it's awfully outdated. Maybe Fedora 30 would be a good goal to shoot for?

**Comment by andrewdavidwong** [comment]
> An update to dom0's Fedora would only make the option available for re-encryption after the fact.

No, I mean a new Qubes release that contains an updated Fedora version in dom0 and requires a clean (re)installation. This would mean that the user performs disk encryption as part of the installation process using an updated installer.

> Here are the defaults for the up-to-date Fedora qube with Argon algs (cryptsetup --help):
> [...]
> I believe Argon2id is widely considered superior and that it should be chosen as a default for the encryption upon installation of Qubes.
> 
> Also note that read the manual it would seem cryptsetup only supports Argon2i and Argon2id, so one of them would have to be chosen.

If I understand correctly, you're saying that even if dom0 were updated to the latest version of Fedora and a clean installation were performed, the result would still not be what you're advocating here. In that case, it sounds like this should be reported as an upstream issue against `cryptsetup` (or possibly Fedora, if Fedora is choosing non-default `cryptsetup` options). [In general, our policy is to respect upstream choices.](https://www.qubes-os.org/faq/#what-is-qubes-attitude-toward-changing-guest-distros) It's up to @marmarek whether Qubes as a project is willing to modify the upstream defaults in this particular case.

> BTW: Please do upgrade the version of Fedora on dom0 soon. I've noticed on multiple occasions it's awfully outdated.

Have you seen this? https://www.qubes-os.org/doc/supported-versions/#note-on-dom0-and-eol

**Comment by DemiMarie** [comment]
@andrewdavidwong Requiring a clean reinstall is incredibly annoying in many cases, mostly because of the need for secure external boot media.  Would it be possible to allow in-place upgrades?

**Comment by DemiMarie** [comment]
@andrewdavidwong Requiring a clean reinstall is incredibly annoying in many cases, mostly because of the need for secure external boot media.  Would it be possible to allow in-place upgrades?

**Comment by marmarek** [comment]
We do plan to support in-place upgrade within major releases 4.x. But it won't by itself re-encrypt your whole disk with a new key and header format.
